### DISCUSSION TITLE
"where self : T" constraint on interfaces

### DISCUSSION THREAD

**Discussion by fa1ntl1ght** [dotnet/csharplang#5299] [Language Ideas]
There is often a need to ensure that the interface is only implemented by classes derived from specific base class.
Allowing to constrain the interface to require specific base class could ease the use of interfaces by providing access to members of that base class in DIM implementations and reduce the number of type checks..

The way I want this to be achieved is probably by some "self" generic constraint I saw discussed in some LDMs.
Then, we could write:

```cs
class ClassA
{
    public int SomeField;
}

interface SomeInterface where self : ClassA
{
    public void DoSomething() //DIM
    {
        SomeField++; //access to a member of the base class
        Console.WriteLine(SomeField);
    }
}
```
The code above shows the basic idea. We have an interface that specifies its required base class and that base class itself. We could then create another class which would derive from ClassA and implement SomeInterface. Then we could be sure that implementations of SomeInterface are all ClassA's derived classes and we could safely cast any SomeInterface to ClassA. Trying to implement SomeInterface in some other non-ClassA-derived class will result in a compiler error.

What do you think? Was this discussed somewhere before?

**UPD: Real world example:**
Unity uses composition as primary design pattern for applications. The main building blocks of an application are `GameObject`s. `GameObject`s contain some generic information and don't define behavior on their own. Instead, behavior is implemented through `Component`s - classes, whose instances can be attached to `GameObject`s. Both `Component` and `GameObject` classes provide various methods to get various information. For example, it is possible to get a component or a list of all components of a specific type attached to a specific `GameObject` through a call to `GetComponent<T>` or `GetComponents<T>` respectively on a `GameObject` or any of its components. Since components are C# classes, they have inheritance and can implement interfaces. It is common to filter components by their interface, e.g. getting all components in all children `GameObject`s that implement specific interface. And then there is an inconvenience in a need to cast every component to `Behaviour` or `MonoBehavior` or any other base class to get some info about its state or other components of its `GameObject`. These casts are redundant, because it is known for sure that these objects are all components and no other class would implement them.
So, at least by adding `where self : Component` or `where self : Behaviour` to Unity interfaces a lot of casts could be removed and these components could be easily stored in collections like `HashSet<ILayoutElement>` etc, without a need to do any casts later. No more `as MonoBehaviour` or `((MonoBehaviour)comp).something`. Just direct access.

**Comment by Joe4evr** [comment]
#5126 #5228

**Reply by fa1ntl1ght** [reply]
As far as I understand, these proposals are about constraining some generic parameter to "self", rather than "self" to something.

**Reply by svick** [reply]
I think the proposal here is a bit different. Though the `SomeInterface` in this proposal would be roughly equivalent to the following code using the other proposals:

```c#
interface SomeInterface<T> where T : self, ClassA
{
    public void DoSomething() //DIM
    {
        SomeField++; //access to a member of the base class
        Console.WriteLine(SomeField);
    }
}
```

**Reply by fa1ntl1ght** [reply]
@svick What proposal is used in your code to access SomeField directly? Isn't casting from `this` to `ClassA` needed here?

**Reply by svick** [reply]
@0x0737 There is no fleshed-out proposal for this, so it's just completely hypothetical code. But I think explicit cast wouldn't be necessary, since the compiler would understand that any type implementing the interface has to inherit from `ClassA`.

**Reply by fa1ntl1ght** [reply]
Ok, thanks, I think I got your idea. So, the compiler will analyse the parameters and will treat a parameter constrained to self as base type of the interface?
The only downside I see here is the verbosity related to generic parameter

**Comment by Richiban** [comment]
This is a really weird proposal from an OO perspective; you're essentially asking that an interface be able to inherit from a class. Think about it, what's different between 

```cs
public class Bar {} 

public interface IFoo where Self : Bar
{
    void M()
    {
        // Use the members of Bar here
    }
}
```

and 

```cs
public class Bar {}

public interface IFoo : Bar
{
    void M()
    {
        // Use the members of Bar here
    }
}
```

**Reply by fa1ntl1ght** [reply]
It's not about interfaces inheriting from classes, it's about only allowing classes derived from certain class to implement certain interface

**Reply by Richiban** [reply]
I understand, but my point is that's kind of the same as asking the interface to inherit from a class.

**Reply by fa1ntl1ght** [reply]
Yes, there are some similarities. But it is not the same as inheriting from class, since the interface is above the class in the hierarchy. That's the reason I didn't use that inheritance-like syntax

**Reply by Richiban** [reply]
"But it is not the same as inheriting from class, since the interface is above the class in the hierarchy" But by enabling an interface to place inheritance requirements on its implementers you are, in effect, flipping the type hierarchy.

Perhaps you could augment your proposal with a real-world example? Unfortunately, at the moment it looks like you're asking for a feature specifically to enable really bad code design; I've seen it before where programmers try to give base types knowledge of their child classes, such as:

```
class Derived : Base {}

class Base
{
    public void M()
    {
        if (this is Derived)
        {
            // ... Ugh
        }
    }
}
```

and it's always the wrong code design this way.

**Reply by tannergooding** [reply]
@Richiban, I think the issue here is that the syntax proposed makes this confusing.

The real world example is `IAdditionOperators<this TSelf>` or possibly `IAdditionOperators<TSelf> where TSelf : this` where its currently defined as `IAdditionOperators<TSelf> where TSelf : IAdditionOperators<TSelf>`. That is, this is just asking for a simplification of the CRTP (`Curiously Recurring Template Pattern`) in a way that provides other additional benefits because its understood by the language.

**Reply by tannergooding** [reply]
And I just realized I may have misunderstood the proposal myself 😄

**Reply by HaloFour** [reply]
@tannergooding 

Right, this seems less about CRTP scenarios and more about allowing the interface to make certain assumptions about its implementer to the extent that members of that implementer (including fields) can be accessed directly.

**Reply by fa1ntl1ght** [reply]
@tannergooding Yes, this isn't about CRTP. @HaloFour explained my idea.

**Reply by fa1ntl1ght** [reply]
Just found this: https://stackoverflow.com/questions/35133887/how-to-require-that-a-protocol-can-only-be-adopted-by-a-specific-class
Looks similar, maybe this is what I want. It seems that Swift also allows other constraints to the self type, like constraining it to only value types or only reference types and so on

**Reply by Richiban** [reply]
You might want to read down a bit further, where another user points out that:

```swift
protocol AddsMoreCommands: UIViewController {
    // code
}
```

is exactly the same as:

```swift
protocol AddsMoreCommands where Self: UIViewController {
   // code
}
```

hence my point about you asking that an interface can inherit from a class. That's what Swift allows

**Reply by fa1ntl1ght** [reply]
@Richiban From Swift 5 release notes:

> Protocols can now constrain their conforming types to those that subclass a given class. Two equivalent forms are supported:
> ```swift
> protocol MyView: UIView { /*...*/ }
> 
> protocol MyView where Self: UIView { /*...*/ }
> ```
So, no, this isn't inheritance
Inheriting means that something is inherited, but interface is on top of the hierarchy, as I said previously.
Inheritance is definitely an incorrect word here

**Reply by Richiban** [reply]
@0x0737 Perhaps we're arguing over terminology here and I shouldn't have implied that it would be _literal_ inheritance by an interface of a class, but it would _effectively_ be so, which I suspect is why Swift chose to co-opt the inheritance syntax for this feature.

**Reply by Richiban** [reply]
As far as I can tell the only difference between selftype constraints (a) and class-to-interface inheritance (b) is a minor syntactic one, where (a) would still require implementing types to explicitly list the base class as being inherited, like so:

```cs
public class Base {}

public interface IFoo where Self : Base {}

public class FooImpl : Base, IFoo {}
```

whereas (b) would not:

```cs
public class Base {}

public interface IFoo : Base {}

public class FooImpl : IFoo {}
```

And thus I consider them pretty much equivalent (from a code writing point of view).

**Reply by fa1ntl1ght** [reply]
@Richiban Anyway, you understood what I want. I don't want to argue any more, so, let it be inheritance, but I don't think this changes anything

**Reply by fa1ntl1ght** [reply]
@Richiban I find the first syntax better due to higher extensibility. It would work good for `struct` and `class` constraints too

**Comment by VolodymyrBS** [comment]
If you want all implementation of `SomeInterface` to be derived from `ClassA` what the point of using interface in first place?
Why not use just ClassA and put `DoSomething()` in there?

```csharp
class ClassA
{
    public int SomeField;

    public virtual void DoSomething()
    {
        SomeField++;
        Console.WriteLine(SomeField);
    }
} 
```

**Reply by fa1ntl1ght** [reply]
This way all derived classes will inherit these members that are just useless for types that don't implement the functionality that supposed to be provided through the interface.

**Reply by DavidArno** [reply]
You are requesting that C# support multiple inheritance via the back-door here. The reason that DIMs do not allow fields is to enable them to provide a safe version of multiple inheritance as they contain no state. 

Remember "favour composition over inheritance". It is already possible to achieve what you want to do via composition and it'll likely lead to a simpler, cleaner solution, too.

**Reply by fa1ntl1ght** [reply]
@DavidArno Looks like you completely misunderstood my proposal. I am not proposing anything related to DIMs for fields. My proposal is about requiring implementations of the interface to derive from some specific base class and providing a simple way to access that class's members from the interface.

**Reply by DavidArno** [reply]
@0x0737, 
You may wish to rethink your initial example then as the first thing I noticed was that you were using this proposal to allow a DIM to access a field, which is currently not allowed. Beyond doing that, it's really unclear just what you hope to gain from this proposal since DIMs already exist and so you can build up this "plug in" functionality that you describe.

Perhaps you could create an example that doesn't access/modify mutable state to explain why DIMs don't fulfil your need?

**Reply by fa1ntl1ght** [reply]
Accessing/modifying the mutable state of a class is the goal. Adding inheritable mutable state to interfaces is not a goal.
Currently, you can cast an interface to some class in a DIM and modify its state, but you never can be sure that the type you are casting to is the actual underlying type of the interface you are writing a DIM for. This proposal allows to specify a requirement that any implementation of an interface must derive from some class. This means you can safely cast that interface to that class and DIMs can safely access the class's state without a risk of getting invalid cast exceptions. This in turn makes explicit casts from `this` to that class redundant and the proposal removes the need to write `((ClassA)this).SomeMember` to access that class's members. This proposal is about type safety and isn't about adding state to interfaces.

**Reply by HaloFour** [reply]
This use case is already covered by interfaces having required protected members to access underlying state.

See [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfgMwAEsCApsgGYQDGjtAkgGID2/WgG8KtcbQAOyfk3ZMM9GAh7LmMCABsAygghMRtAOaMEAbloBfMRJviatfABZaABWQNd+xgAoAlLQAvAB8jsQAnD7casgaOnpMfmYU1pTkDkS0AkIgPNkidlIeAG7eSipgAJ5eBoG0xITUyWkS5XmCAHTRTLFaNVwhtFX9zalUdJkAwgXkrUSEM62tfIK0LKt1MIwA7lmC/s1LEuv8He6eCb5JhamWQA=)

**Reply by VolodymyrBS** [reply]
Why not just add some Interface for ClassA and make SomeInterface derive from it. And also ClassA implements it.
You will not need any casts because you will be sure that every implementation of SomeInterface will implement IClassA.
```csharp
public interface IClassA
{
    int SomeField { get; set; }
}

public class ClassA : IClassA
{
    public int SomeField { get; set; }
}

public interface SomeInterface : IClassA
{
    public void DoSomething()
    {
        SomeField++;
        Console.WriteLine(SomeField);
    }
}
```

**Reply by fa1ntl1ght** [reply]
@HaloFour This doesn't really solve the problem. You still can't access other properties safely. Look at this example:
```cs
public class ClassA
{
	public int fieldA = 1;
}

public class ClassB : ClassA, ISomeInterface
{
	private int _fieldB = 2;
	
	public int someInterfaceState
	{
		get => _fieldB;
		set => _fieldB = value;
	}
}

public interface ISomeInterface
{
	public int someInterfaceState { get; set; }
	
	public void PrintState()
	{
		Console.WriteLine(((ClassA)this).fieldA); //unsafe cast, a type may exist that implements ISomeInterface, but doesn't derive from ClassA
		Console.WriteLine(someInterfaceState);
	}
}
```
The problem lies in accessing the state of the class, not the interface-specific state.

**Reply by fa1ntl1ght** [reply]
@VolodymyrBS Because ClassA may be a class from a third party assembly
Another layer of inheritance will be required, each member must be duplicated in the interface. This is very complicated compared to a simple constraint.

**Reply by HaloFour** [reply]
> You still can't access other properties safely. Look at this example

IMO, you shouldn't access the members or state of an implementer directly.  You should access the members of the contract.

Your use case has a solution.  It might not be the solution that you prefer, but it is a solution that was added to the language very specifically to address this use case.

**Reply by fa1ntl1ght** [reply]
@HaloFour The proposal is about making deriving from specific type a part of contract.

**Reply by HaloFour** [reply]
@0x0737 

I understand that, and I disagree with it.  The interface is the contract, not the implementer.  The interface should make no assumptions about the implementer, it should negotiate via the members of the contract.
